### PR TITLE
Implement selective geometry preview using shaders

### PR DESCRIPTION
**Pull request by ellensi**
### Purpose

This pull request is a better reimplementation of my 2016 recharge project previously addressed by #6610.

A new toggle button called _**Selective Geometry Preview**_ is added to the top-right corner of the Dynamo canvas, just to the left of the Geometry/Node toggle buttons.

When the mode is enabled, only selected nodes will have their corresponding geometries displayed in full opacity. Geometries that correspond to unselected nodes will be displayed transparent with 10% opacity for meshes and 20% opacity for lines and points.

selective

#### Improvements

NEW implementation | Old implementation
------------ | -------------
Transparency flag is passed to the Dynamo shader (more efficient) | Alpha values were recalculated on C# for each mesh triangle (inefficient)
Lines and points can be displayed transparent with 20% opacity | Lines and points were not made transparent yet
Axes and grids are visible among the transparent geometries | Axes and grids were covered behind the supposedly transparent geometries
New button design and localized tooltip string | Dummy button design

#### Limitations

* Semi-transparent `Plane` geometries when selected may cause some unselected transparent geometries to be rendered wrongly
* Changing the "Show edges" setting while under selective geometry preview mode will reset the transparency. As a workaround, user can just click at any blank space of the node canvas.
* Direct Manipulation functionality cannot work properly behind transparent geometries

### Declarations

TBA

- [ ] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@Benglin @Racel 

### FYIs

@riteshchandawar

### CODE REVIEW

**Comment by ikeough** [comment]
@ellensi This is really cool. The transparent draw-order problem is a known issue. We never implemented efficient transparency sorting to take care of that.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 1]
```diff
@@ -2141,11 +2160,18 @@ public int Compare(Model3D x, Model3D y)
                 return result;
             }
 
-            var transA = (bool) a.GetValue(AttachedProperties.HasTransparencyProperty);
-            var transB = (bool) b.GetValue(AttachedProperties.HasTransparencyProperty);
-            result = transA.CompareTo(transB);
+            // selected items under selective geometry preview mode will have higher precedence
+            var selectedA = AttachedProperties.GetSelectivePreview(a) &&
+                !AttachedProperties.GetShowSelected(a) && !AttachedProperties.IsSpecialRenderPackage(a);
+            var selectedB = AttachedProperties.GetSelectivePreview(a) &&
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: This might be a typo `GetSelectivePreview(a)`

**Comment by riteshchandawar** [comment]
@Racel Please take a look at this and suggest few things

1. Location of Button
2. Name of functionality.

**Comment by Racel** [comment]
@ellensi - What happens when frozen nodes are included in the mix? How do you distinguish between frozen and selected geometry?

**Comment by ellensi** [comment]
@Racel 

*On frozen geometries:* Under this new functionality, frozen geometries will be displayed 50% transparent if selected (that is the normal alpha for frozen geometries) and 5% transparent if unselected (that is 50% of 10%). UX suggestions will be more than welcome.

*On selected geometry color:* I personally prefer the blue overlay to be turned off as I feel that it gives a matt effect that blurs the depth perception of geometries. I'd like to hear others' suggestions for this.

image

*About this tool's name:* "Isolate Selected Geometries", or any better idea?

*About a new toggle button:* @riteshchandawar and I have discussed about the current geometry/node toggle button that does not look like a single toggle button. Initially I intended to merge that toggle into something that looks like a single button while adding this new "isolate" button to its left side, but those two new compacted toggle buttons will surely confuse our existing users. While I believe that a settings toolbar will be a good improvement, we're planning to finalize this "isolate" functionality for Dynamo 1.3. How would you finalize the design for now? Should we put it into the Settings menu?

cc: @riteshchandawar @kronz

**Comment by riteshchandawar** [comment]
@Racel During recharge project we put this option in Setting Menu only, but to make to more discoverable, we made it as a button. 

Current Geo/Node buttons are confusing as both doesn't the same job, we should combine those together and only show the other one when the first one is active. For example, When user is in Node View then button will be for changing it to Geometry View and vice versa, so the single button will serve the two purpose. 

But as @ellensi pointed out, if we make that consolidation of existing button and add the new one adjacent to it, then you might get confused between old Geometry button and the new one.

Thoughts?

**Comment by kronz** [comment]
@ellensi @Racel my concern is that we do not add a new UI widget unless we need to.  Seems like the current model for global view settings in the View>Background 3d Preview menu.  

another question:  how does this interact with he Revit preview?  My assumption would be that if you are using this new feature, and Revit Preview is enabled, then the selected elements also show up in Revit.

**Comment by Racel** [comment]
@ellensi @riteshchandawar @kronz - Here are my suggestions:

- Let's get rid of the canvas button for now and put it back in the "Settings Menu". 

- For the name, maybe we should have a setting called "Isolate Selected Geometry"? I could see users always wanting this on or off. So, in a way, it makes sense in the Settings Menu.

- If possible, can you give me a branch build with the button in the canvas? I would like to do some testing on it. Also, if this feature goes into master, will we be able to call this feature in the web UI and test from there if we just add a button?

**Comment by mjkkirschner** [comment]
@Racel  - on the last bullet, I am pretty sure no, since these changes are almost entirely in wpf and in shaders. It will need to be re-implemented in threejs / webGL... if we were to re-implement this at all, or Formit might need to reimplement functionality like this in their viewer if we are to share that going forward?

**Comment by Racel** [comment]
Thanks @mjkkirschner - I see what you mean. If that is the case, the more I think that this feature should live in the "Settings Menu" for now

File: src/DynamoCoreWpf/ViewModels/Watch3D/Shaders/Material.fx [hunk 2]
```diff
@@ -27,7 +27,8 @@ float2 vShadowMapSize = float2(1024, 1024);
 float4 vShadowMapInfo = float4(0.005, 1.0, 0.5, 0.0);
 
 float4 vSelectionColor = float4(0.0, 0.62, 1.0, 1.0);
-
+float4 vTransparentMaterial = 0.1f;
+float4 vTransparentLinePoint = 0.2f;
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/Watch3D/Shaders/Material.fx]: Does this have the same effect as the following?

`float4 vTransparentLinePoint = float4(0.2f, 0.2f, 0.2f, 0.2f);`

File: src/DynamoCoreWpf/ViewModels/Watch3D/DynamoGeometryModel3D.cs [hunk 3]
```diff
@@ -289,6 +289,10 @@ private DynamoMeshVertex[] CreateVertexArray()
             var vertexCount = geometry.Positions.Count;
             var result = new DynamoMeshVertex[vertexCount];
 
+            var isSelected = (bool)GetValue(AttachedProperties.ShowSelectedProperty);
+            var selectiveMode = (bool)GetValue(AttachedProperties.SelectivePreviewProperty);
+            var isSpecialPackage = (bool)GetValue(AttachedProperties.IsSpecialRenderPackageProperty);
+
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/Watch3D/DynamoGeometryModel3D.cs]: Few notes for the above...

1. To align the name with this feature, should `selectiveMode` be `isolationMode`? That may avoid the confusion due to another `isSelected` variable above.
2. Not sure what `isSpecialPackage` mean, can we rename it to `inheritsTransparencyValue` or something like that? All we can do as readers here is to guess its meaning.

File: src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs [hunk 4]
```diff
@@ -2141,11 +2160,18 @@ public int Compare(Model3D x, Model3D y)
...
-            var transA = (bool) a.GetValue(AttachedProperties.HasTransparencyProperty);
-            var transB = (bool) b.GetValue(AttachedProperties.HasTransparencyProperty);
-            result = transA.CompareTo(transB);
+            // selected items under selective geometry preview mode will have higher precedence
+            var selectedA = AttachedProperties.GetSelectivePreview(a) &&
+                !AttachedProperties.GetShowSelected(a) && !AttachedProperties.IsSpecialRenderPackage(a);
+            var selectedB = AttachedProperties.GetSelectivePreview(b) &&
+                !AttachedProperties.GetShowSelected(b) && !AttachedProperties.IsSpecialRenderPackage(b);
+            result = selectedA.CompareTo(selectedB);
+            if (result != 0) return result;
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs]: Please help with documenting these change as discussed, thanks!

**Comment by Benglin** [comment]
All else being addressed, LGTM, thanks!

**Comment by ellensi** [comment]
We merged this in early so that the functionality can be tested in the bug bash. The feature name is set to be "Isolate Selected Geometries" and it is under the Settings menu. Keyboard shortcut (probably Shift+S) will be added in a follow-up pull request. As for now, the selected geometries will stay gray in color. Thanks everyone for your input!

**Pull request merged**